modules/navier_stokes/src/problems/NavierStokesProblem.C

```diff
@@ -0,0 +1,344 @@
+    }
+    else if (_have_mass_matrix) // If we don't have a mass matrix and the user has requested scaling
+                                // then the diagonal of A will be used
+    {
+      mooseAssert(our_parent_Q, "This should be non-null");
+      ierr =
+          MatCreateSubMatrix(our_parent_Q, velocity_is, velocity_is, MAT_INITIAL_MATRIX, &_Q_scale);
+      LIBMESH_CHKERR2(this->comm(), ierr);
+    }
+  }
+  else
+  {
+    mooseAssert(our_parent_Q, "This should be non-null");
+    if (_commute_lsc || _pressure_mass_matrix_as_pre)
+    {
+      ierr =
+          MatCreateSubMatrix(our_parent_Q, pressure_is, pressure_is, MAT_REUSE_MATRIX, &_Q_scale);
+      LIBMESH_CHKERR2(this->comm(), ierr);
+    }
+    else
+    {
+      ierr =
+          MatCreateSubMatrix(our_parent_Q, velocity_is, velocity_is, MAT_REUSE_MATRIX, &_Q_scale);
+      LIBMESH_CHKERR2(this->comm(), ierr);
+    }
+  }
+  ierr = ISDestroy(&pressure_is);
+  LIBMESH_CHKERR2(this->comm(), ierr);
+
+  for (auto & mat : intermediate_Qs)
+  {
+    ierr = MatDestroy(&mat);
+    LIBMESH_CHKERR2(this->comm(), ierr);
+  }
+  for (auto & mat : intermediate_Ls)
+  {
+    ierr = MatDestroy(&mat);
+    LIBMESH_CHKERR2(this->comm(), ierr);
+  }
+
+  ierr = PCFieldSplitGetSubKSP(schur_pc, &num_splits, &subksp);
+  LIBMESH_CHKERR2(this->comm(), ierr);
+  if (num_splits != 2)
+    mooseError("The number of splits should be two");
+  schur_complement_ksp = subksp[1];
+
+  if (_pressure_mass_matrix_as_pre)
+  {
+    mooseAssert(_Q_scale, "This should be non-null");
+    Mat S;
+    ierr = PCFieldSplitSetSchurPre(schur_pc, PC_FIELDSPLIT_SCHUR_PRE_USER, _Q_scale);
+    LIBMESH_CHKERR2(this->comm(), ierr);
+    ierr = KSPGetOperators(schur_complement_ksp, &S, NULL);
+    LIBMESH_CHKERR2(this->comm(), ierr);
+    ierr = KSPSetOperators(schur_complement_ksp, S, _Q_scale);
+    LIBMESH_CHKERR2(this->comm(), ierr);
+  }
+  else // We are doing LSC preconditioning
+  {
+    ierr = KSPGetPC(schur_complement_ksp, &lsc_pc);
+    LIBMESH_CHKERR2(this->comm(), ierr);
+    ierr = PetscObjectTypeCompare(PetscObject(lsc_pc), PCLSC, &is_lsc);
+    LIBMESH_CHKERR2(this->comm(), ierr);
+    if (!is_lsc)
+      mooseError("Not an LSC PC. Please check the 'schur_fs_index' parameter");
+    ierr = PCGetOperators(lsc_pc, NULL, &lsc_pc_pmat);
+    LIBMESH_CHKERR2(this->comm(), ierr);
+
+    if (_commute_lsc)
+    {
+      mooseAssert(_L, "This should be non-null");
+      ierr = PetscObjectCompose((PetscObject)lsc_pc_pmat, "LSC_L", (PetscObject)_L);
+      LIBMESH_CHKERR2(this->comm(), ierr);
+      mooseAssert(_have_mass_matrix, "This is to verify we will enter the next conditional");
+    }
+    if (_have_mass_matrix)
+    {
+      mooseAssert(_Q_scale, "This should be non-null");
+      ierr = PetscObjectCompose((PetscObject)lsc_pc_pmat, "LSC_Qscale", (PetscObject)_Q_scale);
+      LIBMESH_CHKERR2(this->comm(), ierr);
+    }
+  }
+
+  ierr = PetscFree(subksp);
+  LIBMESH_CHKERR2(this->comm(), ierr);
```

Comment:
could use comments throughout

probably should have a reviewer who s more versed in petsc calls. @fdkong @oanaoana ?